### PR TITLE
Alternative mechanism to detect missing `Fabric.backward()` call

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Alternative implementation of #19447 that handles an edge case found by @carmocca. Unfortunately, in this implementation we need to introduce a global variable `_in_fabric_backward` which I tried to avoid before but is unavoidable now. 

I extended the test with the new edge case.

----
📚 Documentation preview 📚: https://pytorch-lightning--19493.org.readthedocs.build/en/19493/

cc @borda @carmocca @justusschock @awaelchli

### CODE REVIEW

File: src/lightning/fabric/wrappers.py [hunk 1]
```diff
@@ -385,3 +388,26 @@ def _capture(model: Any, **kwargs: Any) -> Any:
...
+    from lightning.fabric.fabric import _in_fabric_backward
+
+    # strategy_requires = is_overridden("backward", self._strategy, parent=Strategy)
+    # precision_requires = any(
+    #     is_overridden(method, self._precision, parent=Precision)
+    #     for method in ("pre_backward", "backward", "post_backward")
+    # )
+    # TODO: How do we get a reference to strategy and precision here?
+    strategy_requires = True
+    precision_requires = True
```
Comment (awaelchli) [inline src/lightning/fabric/wrappers.py:403]: Probably have to pass strategy to the FabricModule?

Comment (carmocca) [reply src/lightning/fabric/wrappers.py:402]: In https://github.com/Lightning-AI/pytorch-lightning/pull/18844/files#diff-7cf8c282d7fd9f34b838a3267dcd70a4d7e2434480b656144b631c3cb3620313R106 I had to do it. But then you have to make it a weak reference in the hook in case the tensor is still used when fabric is no longer around.

File: src/lightning/fabric/wrappers.py [hunk 2]
```diff
@@ -385,3 +387,26 @@ def _capture(model: Any, **kwargs: Any) -> Any:
...
+
+    def _backward_hook(_) -> None:
+        if (strategy_requires or precision_requires) and not _in_fabric_backward.get():
+            raise RuntimeError(
+                "The current strategy and precision selection requires you to call `fabric.backward(loss)`"
+                " instead of `loss.backward()`."
+            )
+
+    tensor.register_hook(_backward_hook)
```
Comment (carmocca) [inline src/lightning/fabric/wrappers.py:411]: Does the hook get propagated across operations?

```python
y = model(x)
y.register_hook(backward_hook)
z = y * 2
z = y.sum()
z.backward()  # does z have the same backward hook registerered?
```

Comment (awaelchli) [reply src/lightning/fabric/wrappers.py:411]: The hook is registered on y, because that's what's returned from the module. It will receive a gradient from (y * 2) and fire.

File: src/lightning/fabric/wrappers.py (hunk 1 above)
Comment (awaelchli) [reply src/lightning/fabric/wrappers.py:402]: Yes, that's probably unavoidable.

File: src/lightning/fabric/wrappers.py [hunk 3]
```diff
@@ -123,7 +127,7 @@ def __init__(
         super().__init__()
         self._forward_module = forward_module
         self._original_module = original_module or forward_module
-        self._precision = precision
+        self._strategy = weakref.proxy(strategy)
```
Comment (carmocca) [inline src/lightning/fabric/wrappers.py:130]: I don't think this is right:

The FabricModule should hold a strong reference to the strategy, or else something like

```python
model = MyModel()
model = fabric.setup(model)
fabric = None
model(x)
```

Would fail. It's reasonable to assume that if you have fabric-wrapped the model, you want to keep the strategy alive.

This isn't the same with forward outputs because we don't know what the user will use those tensors for, or when.

Also, I don't think weakref.proxy is correct here. If the reference is not alive by the time it is accessed, it will raise a ReferenceError. We don't want this to ever happen.

I suggest using weakref.ref instead which makes it easier to test if the reference is alive.

Comment (awaelchli) [reply src/lightning/fabric/wrappers.py:130]: Sure I can remove it again. I did it because you asked for this earlier somewhere (or maybe I imagined it).

Comment (carmocca) [reply src/lightning/fabric/wrappers.py:130]: No, we should keep it. I'm saying it should only be weakly held in the backward hook, not the fabric module

Comment (awaelchli) [reply src/lightning/fabric/wrappers.py]: Ok, but I'm not holding a reference to the strategy in the backward hook (maybe I was doing it before when you reviewed). So there should be no issues now?

**Review by carmocca — APPROVED** [review]
Excellent!
